### ISSUE TITLE
raft topology: compactions can make node operations and load balancing progress very slowly

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#14932]
sstable writer captures `erm` since 17d61635481376cdadcfda40f98cb7fdd019fc71, which will hold back the barriers executed as part of topology changes.

Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.

**Comment by mykaul** [comment]
Is this critical for 6.0?

**Comment by kostja** [comment]
We need to define the release criteria for 6.0 first.

**Comment by bhalevy** [comment]
> sstable writer captures `erm` since [17d6163](https://github.com/scylladb/scylladb/commit/17d61635481376cdadcfda40f98cb7fdd019fc71), which will hold back the barriers executed as part of topology changes.
> 
> Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.

Compaction can be stopped for migration.

**Comment by avikivity** [comment]
> > sstable writer captures `erm` since [17d6163](https://github.com/scylladb/scylladb/commit/17d61635481376cdadcfda40f98cb7fdd019fc71), which will hold back the barriers executed as part of topology changes.
> > Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.
> 
> Compaction can be stopped for migration.

No. We're not going to stop all compactions on a cluster every time we want to migrate something.

There's no reason to capture the effective replication map for compaction, and it shouldn't be done for any long-running operations without listening for updates. In this case it's enough to copy the sharder.

**Comment by bhalevy** [comment]
> > > sstable writer captures `erm` since [17d6163](https://github.com/scylladb/scylladb/commit/17d61635481376cdadcfda40f98cb7fdd019fc71), which will hold back the barriers executed as part of topology changes.
> > > Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.
> > 
> > 
> > Compaction can be stopped for migration.
> 
> No. We're not going to stop all compactions on a cluster every time we want to migrate something.

That's not what I meant.  Compactions can be stopped in much finer resolution, like the compaction_group.

> There's no reason to capture the effective replication map for compaction, and it shouldn't be done for any long-running operations without listening for updates. In this case it's enough to copy the sharder.

:+1:

**Comment by avikivity** [comment]
> > > > sstable writer captures `erm` since [17d6163](https://github.com/scylladb/scylladb/commit/17d61635481376cdadcfda40f98cb7fdd019fc71), which will hold back the barriers executed as part of topology changes.
> > > > Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.
> > > 
> > > 
> > > Compaction can be stopped for migration.
> > 
> > 
> > No. We're not going to stop all compactions on a cluster every time we want to migrate something.
> 
> That's not what I meant. Compactions can be stopped in much finer resolution, like the compaction_group.


Any running compaction will hold the effective replication map pointer, it's not enough to stop some, you have to stop all of them. On the entire cluster.

**Comment by bhalevy** [comment]
> > > > > sstable writer captures `erm` since [17d6163](https://github.com/scylladb/scylladb/commit/17d61635481376cdadcfda40f98cb7fdd019fc71), which will hold back the barriers executed as part of topology changes.
> > > > > Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.
> > > > 
> > > > 
> > > > Compaction can be stopped for migration.
> > > 
> > > 
> > > No. We're not going to stop all compactions on a cluster every time we want to migrate something.
> > 
> > 
> > That's not what I meant. Compactions can be stopped in much finer resolution, like the compaction_group.
> 
> Any running compaction will hold the effective replication map pointer, it's not enough to stop some, you have to stop all of them. On the entire cluster.

But with tablet, each table has its own effective replication map.

**Comment by avikivity** [comment]
> > > > > > sstable writer captures `erm` since [17d6163](https://github.com/scylladb/scylladb/commit/17d61635481376cdadcfda40f98cb7fdd019fc71), which will hold back the barriers executed as part of topology changes.
> > > > > > Compactions can take an arbitrary amount of time to complete. This can make topology changes, load balancing in particular, arbitrarily slow.
> > > > > 
> > > > > 
> > > > > Compaction can be stopped for migration.
> > > > 
> > > > 
> > > > No. We're not going to stop all compactions on a cluster every time we want to migrate something.
> > > 
> > > 
> > > That's not what I meant. Compactions can be stopped in much finer resolution, like the compaction_group.
> > 
> > 
> > Any running compaction will hold the effective replication map pointer, it's not enough to stop some, you have to stop all of them. On the entire cluster.
> 
> But with tablets, each table has its own effective replication map.

Ok, so if any tablet in that table compacts, that table will not migrate. And we're not going to stop random compactions for migrations.

effective_replication_map must be held only for short operations, or with a notifier for updates. This is abuse of the API.

**Comment by tgrabiec** [comment]
> Is this critical for 6.0?

I think it is. I also have a patch ready for this. Compaction doesn't really need to hold erm.

**Comment by bhalevy** [comment]
@tgrabiec how is it going with your patch?
Can you submit a PR?
